src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java

```diff
@@ -524,4 +522,58 @@ public void testDefaultConfiguration() throws Exception {
             + "InputCustomImportOrder.java"), expected);
     }
 
+    @Test
+    public void testRulesWithOverlappingPatterns() throws Exception {
+        final DefaultConfiguration checkConfig =
+            createCheckConfig(CustomImportOrderCheck.class);
+        checkConfig.addAttribute("customImportOrderRules",
+                "THIRD_PARTY_PACKAGE###SAME_PACKAGE(6)###STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS");
+        checkConfig.addAttribute("standardPackageRegExp", "com.puppycrawl.tools.*Check$");
+        checkConfig.addAttribute("specialImportsRegExp", "com.puppycrawl.tools.*Tag*");
+        checkConfig.addAttribute("thirdPartyPackageRegExp", "com.puppycrawl.tools.checkstyle.checks.javadoc.*Javadoc*");
+        final String[] expected = {
+            "9: " + getCheckMessage(MSG_ORDER, THIRD, STD, "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocNodeImpl"),
+            "21: " + getCheckMessage(MSG_NONGROUP_EXPECTED, STD, "com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck"),
+            //"24: " + getCheckMessage(MSG_NONGROUP_EXPECTED, SPECIAL, "com.puppycrawl.tools.checkstyle.checks.javadoc.HtmlTag"),
+            "25: " + getCheckMessage(MSG_NONGROUP_EXPECTED, SPECIAL, "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocTag"),
+            "26: " + getCheckMessage(MSG_NONGROUP_EXPECTED, STD, "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocMethodCheck"),
+            "27: " + getCheckMessage(MSG_NONGROUP_EXPECTED, STD, "com.puppycrawl.tools.checkstyle.checks.javadoc.NonEmptyAtclauseDescriptionCheck"),
+            };
+
+        createChecker(checkConfig);
+        verify(checkConfig, getPath("imports" + File.separator
+            + "InputCustomImportOrder_OverlappingPatterns.java"), expected);
+    }
+
+    @Test
+    public void testMultiplePatternMatches() throws Exception {
+        final DefaultConfiguration checkConfig =
+            createCheckConfig(CustomImportOrderCheck.class);
+        checkConfig.addAttribute("customImportOrderRules",
+                "SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE");
+        checkConfig.addAttribute("specialImportsRegExp", "org");
+        checkConfig.addAttribute("standardPackageRegExp", "organ");
+        String[] expected = ArrayUtils.EMPTY_STRING_ARRAY;
+
+        createChecker(checkConfig);
+        verify(checkConfig, new File("src/test/resources-noncompilable/com/puppycrawl/tools/"
+                + "checkstyle/imports/"
+                + "InputCustomImportOrder_MultiplePatternMatches.java").getCanonicalPath(), expected);
+    }
+
+    @Test
+    public void testMultiplePatternMatches2() throws Exception {
+        final DefaultConfiguration checkConfig =
+            createCheckConfig(CustomImportOrderCheck.class);
+        checkConfig.addAttribute("customImportOrderRules",
+                "SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE");
+        checkConfig.addAttribute("specialImportsRegExp", "myO");
```

Comment:
@mkordas changed to ".my". Is it good enough or you mean some scenario that could be closer to real life (it will be more difficult to find one comparing to synthetic test above)?